File: android/app/src/main/kotlin/jp/co/bitz/tictactoe/game/TicTacToePlayer.kt
package jp.co.bitz.tictactoe.game

import jp.co.bitz.gameplaykit.GKGameModelPlayer

/**
 * A fixed X/O player identity, mirroring GameplayKit's `GKGameModelPlayer`. See
 * docs/GAME_DESIGN.md's "Game model" section.
 */
public class TicTacToePlayer private constructor(
    public val mark: Mark,
) : GKGameModelPlayer {
    override val playerId: Int = if (mark == Mark.X) 0 else 1

    public companion object {
        public val x: TicTacToePlayer = TicTacToePlayer(Mark.X)
        public val o: TicTacToePlayer = TicTacToePlayer(Mark.O)

        /** Both players, X then O — [TicTacToeGameModel.players]' fixed value. */
        public val all: List<TicTacToePlayer> = listOf(x, o)

        public fun player(mark: Mark): TicTacToePlayer = if (mark == Mark.X) x else o
    }
}

File: ios/TicTacToe/Game/TicTacToePlayer.swift
import GameplayKit

/// A fixed `X`/`O` player identity, mirroring GameplayKit's `GKGameModelPlayer`. See
/// `docs/GAME_DESIGN.md`'s "Game model" section.
public final class TicTacToePlayer: NSObject, GKGameModelPlayer {
    public let mark: Mark
    public let playerId: Int

    private init(mark: Mark) {
        self.mark = mark
        playerId = mark == .x ? 0 : 1
    }

    public static let x = TicTacToePlayer(mark: .x)
    public static let o = TicTacToePlayer(mark: .o)

    /// Both players, `X` then `O` — `TicTacToeGameModel.players`' fixed value.
    public static let all: [TicTacToePlayer] = [x, o]

    public static func player(for mark: Mark) -> TicTacToePlayer {
        mark == .x ? x : o
    }
}
